core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -0,0 +1,576 @@
+import io.trino.spi.type.BigintType;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.Int128;
+import io.trino.spi.type.Type;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.planner.ExpressionInterpreter;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.SymbolsExtractor;
+import io.trino.sql.planner.TypeAnalyzer;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.AggregationNode.Aggregation;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.tree.Cast;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.NodeRef;
+import io.trino.sql.tree.QualifiedName;
+import io.trino.sql.tree.SearchedCaseExpression;
+import io.trino.sql.tree.SymbolReference;
+import io.trino.sql.tree.WhenClause;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.SystemSessionProperties.isPreAggregateCaseAggregationsEnabled;
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.sql.analyzer.TypeSignatureProvider.fromTypes;
+import static io.trino.sql.analyzer.TypeSignatureTranslator.toSqlType;
+import static io.trino.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static io.trino.sql.planner.plan.AggregationNode.singleGroupingSet;
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+import static io.trino.sql.planner.plan.Patterns.project;
+import static io.trino.sql.planner.plan.Patterns.source;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+import static java.util.function.Predicate.not;
+
+/**
+ * Rule that transforms
+ * <pre> {@code
+ * - Aggregation[key]: sum(aggr1), sum(aggr2), ..., sum(aggrN)
+ *   - Project:
+ *       aggr1 = CASE WHEN col=1 THEN expr ELSE 0
+ *       aggr2 = CASE WHEN col=2 THEN expr ELSE null
+ *       ...
+ *       aggrN = CASE WHEN col=N THEN expr
+ *     - source
+ * }
+ * </pre>
+ * into:
+ * <pre> {@code
+ * - Aggregation[key]: sum(aggr1), sum(aggr2), ..., sum(aggrN)
+ *   - Project:
+ *       aggr1 = CASE WHEN col=1 THEN pre_aggregate
+ *       aggr2 = CASE WHEN col=2 THEN pre_aggregate
+ *       ..
+ *       aggrN = CASE WHEN col=N THEN pre_aggregate
+ *     - Aggregation[key, col]: pre_aggregate = sum(expr)
```

Comment:
Would it make sense to do only partial aggregation here?
The final step for the pre-aggregation may introduce an unnecessary bottleneck, perhaps.

Reply:
- Partial aggregations are not recognized by most of the rules, so this rule would have to be executed near the end of plan optimizations and consider exchanges in pattern too
- Partial aggregation might actually cause this optimization to be less effective as PA doesn't enforce partitioning on `key, col`
- This optimization is a pretty successful heuristic. Without this optimization, input data has a fan-out factor of number of `case` expressions. With this rule, in practice, there won't be any fan-out and usually there is big data reduction with pre-aggregation.
- In the future, PA will be represented by "standard" aggregations too (e.g. intermediate states will be of standard types)